Documentation/Contributors/CodingGuide/README.md

```diff
@@ -0,0 +1,660 @@
+```javascript
+var p = new Cartesian3(1.0, 2.0, 3.0);
+p.x = 'Cesium'; // Changes x to a string, slows down property access
+```
+
+### `from` Constructors
+
+:art: Constructor functions should take the basic components of the class as parameters.  For example, `Cartesian3` takes `x`, `y`, and `z`.
+
+It is often convenient to construct objects from other parameters.  Since JavaScript doesn't have function overloading, Cesium uses static
+functions prefixed with `from` to construct objects in this way.  For example:
+```javascript
+var p = Cartesian3.fromRadians(-2.007, 0.645); // Construct a Cartesian3 object using longitude and latitude
+```
+These are implemented with an optional `result` parameter, which allows callers to pass in a scratch variable:
+```javascript
+Cartesian3.fromRadians = function(longitude, latitude, height, result) {
+    // Compute x, y, z using longitude, latitude, height
+
+    if (!defined(result)) {
+        result = new Cartesian3();
+    }
+
+    result.x = x;
+    result.y = y;
+    result.z = z;
+    return result;
+};
+```
+Since calling a `from` constructor should not require an existing object, the function is assigned to `Cartesian3.fromRadians`, not `Cartesian3.prototype.fromRadians`.
+
+### `to` Functions
+
+Functions that start with `to` return a new type of object, e.g.,
+```javascript
+Cartesian3.prototype.toString = function() {
+    return '(' + this.x + ', ' + this.y + ', ' + this.z + ')';
+};
+```
+
+### Use Prototype Functions for Fundamental Classes Sparingly
+
+:art: Fundamental math classes such as `Cartesian3`, `Quaternion`, `Matrix4`, and `JulianDate` use prototype functions sparingly.  For example, `Cartesian3` does not have a prototype `add` function like this:
+```javascript
+v0.add(v1, result);
+```
+Instead, this is written as
+```javascript
+Cartesian3.add(v0, v1, result);
+```
+The only exceptions are:
+* `clone`
+* `equals`
+* `equalsEpsilon`
+* `toString`
+
+These prototype functions generally delegate to the non-prototype (static) version, e.g.,
+```javascript
+Cartesian3.equals = function(left, right) {
+        return (left === right) ||
+          ((defined(left)) &&
+           (defined(right)) &&
+           (left.x === right.x) &&
+           (left.y === right.y) &&
+           (left.z === right.z));
+};
+
+Cartesian3.prototype.equals = function(right) {
+    return Cartesian3.equals(this, right);
+};
+```
+The prototype versions have the benefit of being able to be used polymorphically.
+
+### Constants
+
+To create a static constant related to a class, use `freezeObject`:
+```javascript
+Cartesian3.ZERO = freezeObject(new Cartesian3(0.0, 0.0, 0.0));
+```
+
+### Private Functions
+
+Like private properties, private functions start with an `_`.  In practice, these are rarely used.  Instead, for better encapsulation, a file-scoped function that takes `this` as the first parameter is used.  For example,
```

Comment:
One typical situation where private instance functions are useful is with our events, which accept function/scope pairs, in order to avoid allocating a closure just to bind `this`.  Perhaps that's too esoteric to bother mentioning here.